### ISSUE TITLE
19.03.9 goroutine stack exceeds 1000000000-byte limit

### ISSUE THREAD

**Issue by Antiarchitect** [moby/moby#41044]
**Description**
Complicated multi-stage build with multiple --cache-from and COPY --from <previous-stage>
Logs: https://gist.github.com/Antiarchitect/583d7a75f7cbc5f19ebff9b9f38c8d0a

**Steps to reproduce the issue:**
1. Run the build
2. Wait for docker daemon crash

**Describe the results you received:**
Docker daemon crashes, build log ends with unexpected EOF

**Describe the results you expected:**
Build passes normally

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           19.03.9
 API version:       1.40
 Go version:        go1.13.10
 Git commit:        9d98839
 Built:             Fri May 15 00:26:15 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.9
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.10
  Git commit:       9d98839
  Built:            Fri May 15 00:24:15 2020
  OS/Arch:          linux/amd64
  Experimental:     true
 containerd:
  Version:          1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Output of `docker info`:**

```
Client:                                                                                                                                                                                           [0/1823]
 Debug Mode: false

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 1
 Server Version: 19.03.9
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.6.13-300.fc32.x86_64
 Operating System: Fedora 32 (Thirty Two)
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 15.57GiB
 Name: localhost.localdomain
 ID: P5EM:M6NC:P7AY:35ET:CJUH:AITQ:CENP:LJB7:4IOL:ZQMC:YVJG:F7OK
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Username: antiarchitect
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: true
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Fedora 32, local build by hand

**Comment by Antiarchitect** [comment]
Found related: https://github.com/moby/moby/issues/40993

**Comment by Antiarchitect** [comment]
Still present in 19.03.10

**Comment by thaJeztah** [comment]
Do you have steps to reproduce? Is this with BuildKit enabled, disabled? Is it only "related to" https://github.com/moby/moby/issues/40993, or the same situation/problem?

**Comment by Antiarchitect** [comment]
DOCKER_BUILDKIT=1
BUILDKIT_INLINE_CACHE=1

And it seems like I have https://github.com/moby/moby/issues/40993#issuecomment-634259286 the same situation with Dockerfile structure. Is it known in which version of Docker buildkit version with this patch: https://github.com/moby/buildkit/pull/1382 will be used?

**Comment by thaJeztah** [comment]
Looks like it's currently in master and BuildKit v0.7.x, but not (yet) in the version used by Docker 19.03 (https://github.com/moby/buildkit/commits/docker-19.03/solver/exporter.go). I can open a backport for consideration by the BuildKit maintainers

**Comment by Antiarchitect** [comment]
@thaJeztah please do :) It would be very nice as I cannot reduce my build complexity with ease.

**Comment by thaJeztah** [comment]
Opened https://github.com/moby/buildkit/pull/1514. After that's merged, it needs to be vendor in the 19.03 branch in this repository to resolve the issue for 19.03

**Comment by Antiarchitect** [comment]
@thaJeztah Offtopic: Is there any transparent way to know which Buildkit version (commit?) a Docker release use? It seems like no info in `docker info`.

**Comment by thaJeztah** [comment]
BuildKit is compiled into the `dockerd` daemon; you can find which version is used in the `vendor.conf` file. For example, for v19.03.9; https://github.com/moby/moby/blob/v19.03.9/vendor.conf#L29

 (I see the v19.03.10 tag is not yet in this repository, only in the [docker/engine downstream](https://github.com/docker/engine/releases/tag/v19.03.10), I'll ask for that to be addressed)

**Comment by cpuguy83** [comment]
In the meantime, you can use `docker buildx` to create a buildkit instance that runs in a container that you can build against.

```
docker buildx create --name my_builder --use
docker buildx build ...
```

The  downside is if you want to import the image to docker, its an extra flag (and time) to do so.

**Comment by Antiarchitect** [comment]
Will this https://github.com/moby/buildkit/pull/1514 be merged in next patch release?

**Comment by thaJeztah** [comment]
Opened https://github.com/moby/moby/pull/41185 to vendor the fix in the 19.03 branch

**Comment by Antiarchitect** [comment]
@thaJeztah Just asking (as 19.03.13 is out) is there any progress on this? Did this included?

**Comment by thaJeztah** [comment]
yes; the fix mentioned above is included in docker 19.03.13

**Comment by Antiarchitect** [comment]
@thaJeztah Closing. Thank you and all of your team for the hard work!
